### PR TITLE
PLT-6233: Fix Flat (Binder FakeNamedDeBruijn) instance. Add tests

### PR DESCRIPTION
**Pull request by bezirg**
The bug is that Flat (Binder FakeNamedDeBruijn) was not roundtripping. 

The encode and size methods in that instance were wrong: they were supposed to mirror the methods of `Binder DeBruijn`, but they were instead mirroring the methods of `Binder NamedDeBruijn`.

These two problematic methods are not exercised "in the wild",
because people do not encode through FND (and Binder FND) but directly through D (and Binder D). See https://github.com/input-output-hk/plutus/blob/master/plutus-ledger-api/src/PlutusLedgerApi/Common/SerialisedScript.hs#L82-L98

The decode method was not wrong and remains ghc-core-wise the same.
The decode method is the most important, because that is exercised in the wild.

The provided tests reveal the bug on master.

Pre-submit checklist:
- Branch
    - [x] Tests are provided (if possible)
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [ ] Changelog fragments have been written (if appropriate)
    - [x] Relevant tickets are mentioned in commit messages
    - [x] Formatting, PNG optimization, etc. are updated
- PR
    - [ ] (For external contributions) Corresponding issue exists and is linked in the description
    - [x] Targeting master unless this is a cherry-pick backport
    - [ ] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested

### CODE REVIEW

File: plutus-core/untyped-plutus-core/test/Flat/Spec.hs [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+deriving via Word64 instance Arbitrary DeBruijn
+instance Arbitrary FakeNamedDeBruijn where
+    arbitrary= toFake <$> arbitrary -- via debruijn
+deriving newtype instance Arbitrary (Binder DeBruijn)
+deriving newtype instance Arbitrary (Binder FakeNamedDeBruijn)
+
+test_iso :: TestTree
+test_iso = testProperty "fakeIso" $ \d fnd ->
+         (d === fromFake (toFake d))
+    .&&. (fnd === toFake (fromFake fnd))
```
Comment (effectfully) [inline plutus-core/untyped-plutus-core/test/Flat/Spec.hs:30]: These look like two distinct properties, why merge them into a single one? If it ever fails, somebody will need to spend time figuring out which one has failed. Plus you get worse coverage as `d` staying constant and `fnd` changing is a considered a different test case, even though it's not different for `d`.

File: plutus-core/untyped-plutus-core/test/Flat/Spec.hs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+test_deBruijnTripping :: TestTree
+test_deBruijnTripping = testProperty "debruijnTripping" $ \(d :: DeBruijn) (f :: FakeNamedDeBruijn) ->
+      Right d === (unflat $ flat d)
+  .&&. Right f === (unflat $ flat f)
```
Comment (effectfully) [inline plutus-core/untyped-plutus-core/test/Flat/Spec.hs:35]: Same.

File: plutus-core/untyped-plutus-core/test/Flat/Spec.hs [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+test_binderDecode :: TestTree
+test_binderDecode = testProperty "binder" $ \(b :: Binder DeBruijn) (bf :: Binder FakeNamedDeBruijn) ->
+   -- binders should always decode as init binder
+   Right initB === (unflat $ flat b)
+ .&&. Right (toFake <$> initB) === (unflat $ flat bf)
```
Comment (effectfully) [inline plutus-core/untyped-plutus-core/test/Flat/Spec.hs:41]: Same?

File: plutus-core/plutus-core/src/PlutusCore/Flat.hs [hunk 4]
```diff
@@ -404,16 +403,23 @@ deriving newtype instance Flat (Binder TyName)
...
+See NOTE: [Why newtype FakeNamedDeBruijn]
+NOTE: the serialization roundtrip holds iff the invariant binder.index==0 holds
+-}
 instance Flat (Binder FakeNamedDeBruijn) where
-    size  = size . fromFake . coerce -- via binder debruijn
-    encode = encode . fromFake . coerce -- via binder debruijn
-    decode = coerce . toFake . coerce <$> decode @(Binder DeBruijn) -- via binder debruijn
+    size  = size . fmap fromFake
+    encode = encode . fmap fromFake
+    decode =  fmap toFake <$> decode
```
Comment (effectfully) [inline plutus-core/plutus-core/src/PlutusCore/Flat.hs:425]: Is performance important here? We can do it better performance-wise.

**Comment by bezirg** [comment]
@effectfully not done pushing yet, please review when undrafted

**Review by effectfully — COMMENTED** [review]
This is a prereview, I'm going to try to understand what problem this PR fixes and then perform an actual review.

**Comment by effectfully** [comment]
> @effectfully not done pushing yet, please review when undrafted

Oops, sorry!

Could you please add a description of the bug somewhere? I couldn't find it in Jira or here.

File: plutus-core/plutus-core/src/PlutusCore/Flat.hs (hunk 4 above)
Comment (bezirg) [reply plutus-core/plutus-core/src/PlutusCore/Flat.hs:425]: I don't think so. But good that u mention it. I wante do to check if there is any ghc-core difference in the decode path.

**Comment by bezirg** [comment]
> > @effectfully not done pushing yet, please review when undrafted
> 
> Oops, sorry!
> 
> Could you please add a description of the bug somewhere? I couldn't find it in Jira or here.

The bus is that Flat (Binder FakeNamedDeBruijn) was not roundtripping. The encode and size functions in that instance were wrong. The bug does not happen "in the wild", because people do not encode through FND but directly through D

File: plutus-core/plutus-core/src/PlutusCore/Flat.hs (hunk 4 above)
Comment (effectfully) [reply plutus-core/plutus-core/src/PlutusCore/Flat.hs:425]: You can probably use `PlutusPrelude.coerceVia (. fmap fromFake) size` and the same for the rest. It's almost as readable, but is definitely faster, although probably negligibly so. But even then I believe it's worth it.

Comment (bezirg) [reply plutus-core/plutus-core/src/PlutusCore/Flat.hs:425]: I see what you are suggesting, but `Coercible FakeNamedDeBruijn DeBruijn` does not hold. They are isomorphic but they do not have the same runtime representation. So I don't see how I can use the Coercible.coerce mechanism.

Comment (effectfully) [reply plutus-core/plutus-core/src/PlutusCore/Flat.hs:425]: Oh I see, I just looked at the code and now I understand.

OK, could you please post what GHC Core this instance compiles to?

Comment (bezirg) [reply plutus-core/plutus-core/src/PlutusCore/Flat.hs:427]: It is very difficult to read the ghc core because I don't get good names for the dictionaries. I was expecting something like $fFlatBinderDeBruijn  and $fFlatBinderFakeNamedDeBruijn but I am getting $fflatBinder + a digit.
this is for all the "Flexible Instances" , e.g. instance Flat (Binder DB) , instance Flat (Binder FND) , instance Flat (Binder NDB).. etc.

But I am fairly sure that the decode method implementation for `Flat (Binder FakeNamedDeBruijn)` is this one, and has remained exactly the same:


```
-- RHS size: {terms: 17, types: 23, coercions: 3, joins: 0/0}
$fFlatBinder7
  = \ _ s s1 ->
      case s of { S dt1 dt2 ->
      (# s1,
         case fakeName of dt3 { Text ipv ipv1 ipv2 ->
         GetResult dt1 dt2 ((NamedDeBruijn dt3 0##) `cast` <Co:3>)
         } #)
      }
```

I will send you the whole diff by email.

**Comment by bezirg** [comment]
Still it needs a changelog and to look at golden output more closely (perhaps I have some golden leftovers after renaming), but is in a good state for review @effectfully

**Comment by bezirg** [comment]
> Comparing benchmark results of ' plutus-benchmark:validation-decode' on 'a1b9bb57d5' (base) and 'f3d4b2be55' (PR)
> 
> Results table

Strange results again. I was expecting 0% difference

File: plutus-core/plutus-core/src/PlutusCore/Flat.hs (hunk 4 above)
Comment (bezirg) [reply plutus-core/plutus-core/src/PlutusCore/Flat.hs:427]: Yes, decoding only. I dnt know how to improve it. Maybe NOINLINE can help?

**Comment by bezirg** [comment]
Golden files are ok, it was just a git renaming switcheroo that led me to believe they are wrong

**Comment by kwxm** [comment]
I ran the benchmarks manually on the benchmarking machine and got a uniform improvement.  I'll do it again to see what happens.  Nothing else is running on the machine.

```
Script                              Old            New         Change
----------------------------------------------------------------------------------
auction_1-1                       184.5 μs	 181.0 μs      -1.9%
auction_1-2                       735.8 μs	 720.5 μs      -2.1%
auction_1-3                       726.8 μs	 711.9 μs      -2.1%
auction_1-4                       236.4 μs	 232.8 μs      -1.5%
auction_2-1                       186.6 μs	 182.7 μs      -2.1%
auction_2-2                       737.0 μs	 721.2 μs      -2.1%
auction_2-3                       969.5 μs	 941.8 μs      -2.9%
auction_2-4                       726.6 μs	 709.9 μs      -2.3%
auction_2-5                       236.7 μs	 232.4 μs      -1.8%
crowdfunding-success-1            220.5 μs	 214.2 μs      -2.9%
crowdfunding-success-2            219.9 μs	 213.6 μs      -2.9%
crowdfunding-success-3            220.1 μs	 212.8 μs      -3.3%
currency-1                        271.0 μs	 263.5 μs      -2.8%
escrow-redeem_1-1                 382.3 μs	 375.2 μs      -1.9%
escrow-redeem_1-2                 382.8 μs	 375.1 μs      -2.0%
escrow-redeem_2-1                 447.2 μs	 436.7 μs      -2.3%
escrow-redeem_2-2                 445.2 μs	 435.2 μs      -2.2%
escrow-redeem_2-3                 448.3 μs	 435.5 μs      -2.9%
escrow-refund-1                   160.9 μs	 157.0 μs      -2.4%
future-increase-margin-1          271.5 μs	 263.9 μs      -2.8%
future-increase-margin-2          593.4 μs	 582.6 μs      -1.8%
future-increase-margin-3          594.7 μs	 585.1 μs      -1.6%
future-increase-margin-4          551.7 μs	 542.7 μs      -1.6%
future-increase-margin-5          940.4 μs	 917.9 μs      -2.4%
future-pay-out-1                  270.9 μs	 263.0 μs      -2.9%
future-pay-out-2                  592.8 μs	 586.3 μs      -1.1%
future-pay-out-3                  597.0 μs	 585.8 μs      -1.9%
future-pay-out-4                  939.0 μs	 913.0 μs      -2.8%
future-settle-early-1             271.2 μs	 263.2 μs      -2.9%
future-settle-early-2             592.8 μs	 584.4 μs      -1.4%
future-settle-early-3             594.8 μs	 582.4 μs      -2.1%
future-settle-early-4             698.8 μs	 683.9 μs      -2.1%
game-sm-success_1-1               435.3 μs	 425.1 μs      -2.3%
game-sm-success_1-2               202.7 μs	 200.4 μs      -1.1%
game-sm-success_1-3               725.2 μs	 707.3 μs      -2.5%
game-sm-success_1-4               236.0 μs	 234.4 μs      -0.7%
game-sm-success_2-1               431.3 μs	 420.3 μs      -2.6%
game-sm-success_2-2               202.4 μs	 201.2 μs      -0.6%
game-sm-success_2-3               725.3 μs	 707.6 μs      -2.4%
game-sm-success_2-4               235.8 μs	 233.2 μs      -1.1%
game-sm-success_2-5               721.2 μs	 704.7 μs      -2.3%
game-sm-success_2-6               236.2 μs	 233.5 μs      -1.1%
multisig-sm-1                     444.6 μs	 434.6 μs      -2.2%
multisig-sm-2                     432.6 μs	 425.6 μs      -1.6%
multisig-sm-3                     437.0 μs	 430.7 μs      -1.4%
multisig-sm-4                     444.0 μs	 434.6 μs      -2.1%
multisig-sm-5                     637.3 μs	 622.9 μs      -2.3%
multisig-sm-6                     445.7 μs	 434.1 μs      -2.6%
multisig-sm-7                     433.3 μs	 426.4 μs      -1.6%
multisig-sm-8                     439.2 μs	 430.4 μs      -2.0%
multisig-sm-9                     441.7 μs	 435.4 μs      -1.4%
multisig-sm-10                    636.1 μs	 622.1 μs      -2.2%
ping-pong-1                       368.3 μs	 359.3 μs      -2.4%
ping-pong-2                       368.5 μs	 361.2 μs      -2.0%
ping-pong_2-1                     213.7 μs	 211.5 μs      -1.0%
prism-1                           171.6 μs	 169.1 μs      -1.5%
prism-2                           455.3 μs	 446.2 μs      -2.0%
prism-3                           403.5 μs	 394.0 μs      -2.4%
pubkey-1                          144.7 μs	 143.0 μs      -1.2%
stablecoin_1-1                    1.032 ms	 1.016 ms      -1.6%
stablecoin_1-2                    197.7 μs	 195.7 μs      -1.0%
stablecoin_1-3                    1.188 ms	 1.167 ms      -1.8%
stablecoin_1-4                    210.8 μs	 208.8 μs      -0.9%
stablecoin_1-5                    1.518 ms	 1.490 ms      -1.8%
stablecoin_1-6                    260.7 μs	 258.2 μs      -1.0%
stablecoin_2-1                    1.030 ms	 1.016 ms      -1.4%
stablecoin_2-2                    198.3 μs	 196.5 μs      -0.9%
stablecoin_2-3                    1.187 ms	 1.166 ms      -1.8%
stablecoin_2-4                    211.4 μs	 208.3 μs      -1.5%
token-account-1                   201.6 μs	 197.8 μs      -1.9%
token-account-2                   368.5 μs	 360.2 μs      -2.3%
uniswap-1                         457.7 μs	 443.8 μs      -3.0%
uniswap-2                         236.0 μs	 231.0 μs      -2.1%
uniswap-3                         1.994 ms	 1.939 ms      -2.8%
uniswap-4                         343.5 μs	 335.1 μs      -2.4%
uniswap-5                         1.283 ms	 1.258 ms      -1.9%
uniswap-6                         328.0 μs	 321.8 μs      -1.9%
vesting-1                         394.0 μs	 385.1 μs      -2.3%
```

**Comment by kwxm** [comment]
OK, this is hopeless.  I ran the benchmarks in both branches again and they're not even consistent in the same branch.

File: plutus-core/plutus-core/src/PlutusCore/DeBruijn/Internal.hs [hunk 5]
```diff
@@ -80,8 +115,14 @@ data NamedDeBruijn = NamedDeBruijn {ndbnString :: !T.Text, ndbnIndex :: !Index}
   deriving stock (Show, Generic, Read)
   deriving anyclass (NFData)
 
--- | A wrapper around nameddebruijn that must hold the invariant of name=`fakeName`.
-newtype FakeNamedDeBruijn = FakeNamedDeBruijn NamedDeBruijn
+{-| A wrapper around `NamedDeBruijn` that *must* hold the invariant of name=`fakeName`.
+
+We do not export the `FakeNamedDeBruijn` constructor: the projection `FND->ND` is safe
+but injection `ND->FND` is unsafe, thus they are not isomorphic.
```
Comment (kwxm) [inline plutus-core/plutus-core/src/PlutusCore/DeBruijn/Internal.hs:121]: I don't know if "unsafe" is exactly the right word here, but I don't have anything better to suggest.

**Review by kwxm — APPROVED** [review]
This looks fine: most of my comments are about typos and stuff.  I'm not sure that I understand all the subtleties related to coercion though.

**Comment by kwxm** [comment]
I tried benchmarking this on the benchmark machine and had trouble getting sensible results.  I gave up and tried on an old (and slow) desktop machine which seems to give quite consistent results.  That did seem to confirm that this PR introduces a slowdown in the `validation-decode` benchmarks:
```
Script                              Old            New         Change
----------------------------------------------------------------------------------
auction_1-1                       381.7 μs	 400.9 μs      +5.0%
auction_1-2                       1.090 ms	 1.126 ms      +3.3%
auction_1-3                       1.089 ms	 1.140 ms      +4.7%
auction_1-4                       381.7 μs	 404.6 μs      +6.0%
auction_2-1                       382.2 μs	 401.8 μs      +5.1%
auction_2-2                       1.083 ms	 1.146 ms      +5.8%
auction_2-3                       1.091 ms	 1.152 ms      +5.6%
auction_2-4                       1.092 ms	 1.133 ms      +3.8%
auction_2-5                       381.9 μs	 406.1 μs      +6.3%
crowdfunding-success-1            469.0 μs	 488.7 μs      +4.2%
crowdfunding-success-2            469.5 μs	 488.4 μs      +4.0%
crowdfunding-success-3            472.4 μs	 489.4 μs      +3.6%
currency-1                        463.5 μs	 480.7 μs      +3.7%
escrow-redeem_1-1                 627.9 μs	 653.1 μs      +4.0%
escrow-redeem_1-2                 628.5 μs	 652.5 μs      +3.8%
escrow-redeem_2-1                 627.9 μs	 653.4 μs      +4.1%
escrow-redeem_2-2                 628.2 μs	 653.5 μs      +4.0%
escrow-redeem_2-3                 632.4 μs	 652.1 μs      +3.1%
escrow-refund-1                   628.3 μs	 653.3 μs      +4.0%
future-increase-margin-1          462.9 μs	 481.0 μs      +3.9%
future-increase-margin-2          633.1 μs	 658.1 μs      +3.9%
future-increase-margin-3          633.5 μs	 657.9 μs      +3.9%
future-increase-margin-4          1.431 ms	 1.481 ms      +3.5%
future-increase-margin-5          1.432 ms	 1.495 ms      +4.4%
future-pay-out-1                  462.6 μs	 480.8 μs      +3.9%
future-pay-out-2                  633.7 μs	 658.0 μs      +3.8%
future-pay-out-3                  633.1 μs	 659.2 μs      +4.1%
future-pay-out-4                  1.431 ms	 1.482 ms      +3.6%
future-settle-early-1             462.6 μs	 480.6 μs      +3.9%
future-settle-early-2             633.7 μs	 659.1 μs      +4.0%
future-settle-early-3             634.1 μs	 658.1 μs      +3.8%
future-settle-early-4             1.432 ms	 1.480 ms      +3.4%
game-sm-success_1-1               1.058 ms	 1.095 ms      +3.5%
game-sm-success_1-2               326.8 μs	 342.6 μs      +4.8%
game-sm-success_1-3               1.058 ms	 1.096 ms      +3.6%
game-sm-success_1-4               327.2 μs	 342.7 μs      +4.7%
game-sm-success_2-1               1.057 ms	 1.101 ms      +4.2%
game-sm-success_2-2               326.6 μs	 342.4 μs      +4.8%
game-sm-success_2-3               1.056 ms	 1.095 ms      +3.7%
game-sm-success_2-4               327.4 μs	 342.8 μs      +4.7%
game-sm-success_2-5               1.057 ms	 1.095 ms      +3.6%
game-sm-success_2-6               327.1 μs	 342.7 μs      +4.8%
multisig-sm-1                     1.179 ms	 1.220 ms      +3.5%
multisig-sm-2                     1.177 ms	 1.219 ms      +3.6%
multisig-sm-3                     1.177 ms	 1.221 ms      +3.7%
multisig-sm-4                     1.178 ms	 1.219 ms      +3.5%
multisig-sm-5                     1.179 ms	 1.220 ms      +3.5%
multisig-sm-6                     1.180 ms	 1.218 ms      +3.2%
multisig-sm-7                     1.176 ms	 1.218 ms      +3.6%
multisig-sm-8                     1.176 ms	 1.220 ms      +3.7%
multisig-sm-9                     1.177 ms	 1.217 ms      +3.4%
multisig-sm-10                    1.179 ms	 1.217 ms      +3.2%
ping-pong-1                       977.4 μs	 1.014 ms      +3.7%
ping-pong-2                       976.6 μs	 1.014 ms      +3.8%
ping-pong_2-1                     978.6 μs	 1.013 ms      +3.5%
prism-1                           318.0 μs	 333.8 μs      +5.0%
prism-2                           1.040 ms	 1.076 ms      +3.5%
prism-3                           465.6 μs	 485.6 μs      +4.3%
pubkey-1                          326.0 μs	 342.4 μs      +5.0%
stablecoin_1-1                    2.097 ms	 2.156 ms      +2.8%
stablecoin_1-2                    327.1 μs	 342.5 μs      +4.7%
stablecoin_1-3                    2.100 ms	 2.155 ms      +2.6%
stablecoin_1-4                    327.0 μs	 342.8 μs      +4.8%
stablecoin_1-5                    2.098 ms	 2.157 ms      +2.8%
stablecoin_1-6                    326.9 μs	 342.9 μs      +4.9%
stablecoin_2-1                    2.099 ms	 2.158 ms      +2.8%
stablecoin_2-2                    327.0 μs	 343.6 μs      +5.1%
stablecoin_2-3                    2.184 ms	 2.154 ms      -1.4%
stablecoin_2-4                    326.6 μs	 342.8 μs      +5.0%
token-account-1                   460.6 μs	 478.7 μs      +3.9%
token-account-2                   430.3 μs	 444.1 μs      +3.2%
uniswap-1                         485.9 μs	 485.5 μs      -0.1%
uniswap-2                         460.2 μs	 478.5 μs      +4.0%
uniswap-3                         1.517 ms	 1.569 ms      +3.4%
uniswap-4                         353.0 μs	 369.4 μs      +4.6%
uniswap-5                         1.516 ms	 1.569 ms      +3.5%
uniswap-6                         353.0 μs	 368.9 μs      +4.5%
vesting-1                         641.9 μs	 666.0 μs      +3.8%
```
I ran the benchmarks several times and got similar results each time.

**Comment by bezirg** [comment]
thank you @kwxm ! I cannot explain this real slowdown unfortunately. The core looks the same to me at the affected place.

File: plutus-core/plutus-core/src/PlutusCore/DeBruijn.hs [hunk 6]
```diff
@@ -6,10 +6,14 @@
...
     ( Index (..)
+    , Level (..)
+    , LevelInfo (..)
     , HasIndex (..)
     , DeBruijn (..)
     , NamedDeBruijn (..)
-    , FakeNamedDeBruijn (..)
+    -- we follow the same approach as Renamed, expose the constructor from Internal module,
+    -- but hide it in this parent module.
+    , FakeNamedDeBruijn (unFakeNamedDeBruijn)
```
Comment (bezirg) [inline plutus-core/plutus-core/src/PlutusCore/DeBruijn.hs:16]: actually exposing the field is unsafe because it can be used as a constructor, but we can address this (and other such instances) in another PR

File: plutus-core/plutus-core/src/PlutusCore/Core/Type.hs [hunk 7]
```diff
@@ -246,6 +246,7 @@ mapFun f = go where
 -- This marking allows us to skip the (de)serialization of binders at LamAbs/TyAbs positions
 -- iff 'name' is DeBruijn-encoded (level or index). See for example the instance of  'UntypedPlutusCore.Core.Instance.Flat'
 newtype Binder name = Binder { unBinder :: name }
+    deriving stock (Eq, Show, Functor)
```
Comment (effectfully) [inline plutus-core/plutus-core/src/PlutusCore/Core/Type.hs:249]: It's not really important, but please make it

```haskell
newtype Binder name = Binder { unBinder :: name }
    deriving stock (Eq, Show)

instance Functor Binder where
    fmap = coerce
    {-# INLINE fmap #-}
```

because apparently GHC generates unoptimal code (see the slack [discussion](https://input-output-rnd.slack.com/archives/C48HA39RS/p1687917638566799)).

File: plutus-core/plutus-core/stdlib/PlutusCore/StdLib/Data/Bool.hs [hunk 8]
```diff
@@ -22,15 +22,15 @@ import PlutusCore.StdLib.Data.Unit
 import Universe
 
 -- | 'Bool' as a PLC type.
-bool :: uni `Includes` Bool => Type TyName uni ()
+bool :: uni `Includes` Bool => Type tyname uni ()
```
Comment (effectfully) [inline plutus-core/plutus-core/stdlib/PlutusCore/StdLib/Data/Bool.hs:25]: I think you just introduced a whole lot of git conflicts for my half-ready PR... Why the change?

File: plutus-core/prelude/PlutusPrelude.hs [hunk 9]
```diff
@@ -240,3 +244,7 @@ zipExact _ _           = Nothing
 unsafeFromRight :: (Show e) => Either e a -> a
 unsafeFromRight (Right a) = a
 unsafeFromRight (Left e)  = error $ show e
+
+-- | function recursively applied N times
+timesA :: Natural -> (a -> a) -> a -> a
+timesA = ala Endo . stimes
```
Comment (effectfully) [inline plutus-core/prelude/PlutusPrelude.hs:250]: I've done some benchmarking a while ago and it was actually not a good way to define this function. The issue is that in

```haskell
instance Semigroup (Endo a) where
        (<>) = coerce ((.) :: (a -> a) -> (a -> a) -> (a -> a))
        stimes = stimesMonoid
```

`stimesMonoid` is defined like this:

```haskell
stimesMonoid :: (Integral b, Monoid a) => b -> a -> a
stimesMonoid n x0 = case compare n 0 of
  LT -> errorWithoutStackTrace "stimesMonoid: negative multiplier"
  EQ -> mempty
  GT -> f x0 n
    where
      f x y
        | even y = f (x `mappend` x) (y `quot` 2)
        | y == 1 = x
        | otherwise = g (x `mappend` x) (y `quot` 2) x               -- See Note [Half of y - 1]
      g x y z
        | even y = g (x `mappend` x) (y `quot` 2) z
        | y == 1 = x `mappend` z
        | otherwise = g (x `mappend` x) (y `quot` 2) (x `mappend` z) -- See Note [Half of y - 1]
```

which has quite a lot of bookkeeping that doesn't seem to benefit simple function composition at all. Not sure why they do it this way for function composition.

File: plutus-core/untyped-plutus-core/test/DeBruijn/UnDeBruijnify.hs [hunk 10]
```diff
@@ -1,159 +1,88 @@
...
-        . flip evalStateT mempty
-        . progTerm (unDeBruijnTermWith freeIndexAsConsistentLevel) . mkProg
-    mkProg = UPLC.Program () PLC.latestVersion . termMapNames fakeNameDeBruijn
+    nestedGoldenVsPretty act (n,t) =
+        nestedGoldenVsDoc n ".uplc" $ toPretty $ act $ mkProg t
+
+    actDefault = progTerm unDeBruijnTerm
+
+    actGrace = progTerm (unDeBruijnTermWith freeIndexAsConsistentLevel)
+               >>> flip evalStateT mempty
```
Comment (effectfully) [inline plutus-core/untyped-plutus-core/test/DeBruijn/UnDeBruijnify.hs:83]: meh

**Review by effectfully — APPROVED** [review]
I don't think you've introduced any slowdown with this PR, I'm going to conclude that it's our benchmarks that are unreliable, because as you say the Core is identical where it matters.

I do think we can make this whole thing faster, but then I also think we should try removing `FakeNamedDeBruijn` and all related logic entirely and not do anything with names on the chain. I'll make a ticket.

Don't appreciate the future git conflicts!

This does fix a bug, so it would be great to get it merged.

File: plutus-core/plutus-core/stdlib/PlutusCore/StdLib/Data/Bool.hs (hunk 8 above)
Comment (bezirg) [reply plutus-core/plutus-core/stdlib/PlutusCore/StdLib/Data/Bool.hs:25]: These changes in stdlib are contained in this commit: https://github.com/input-output-hk/plutus/pull/5390/commits/147e2ee78e0ac936e3f1bd12558d587058e5896f  . I did this to generalize a bit the stdlib to help me reuse parts of it in my debruijn tests. The stdlib otherwise is not "debruijn friendly" (relies on renaming a lot)

File: plutus-core/prelude/PlutusPrelude.hs (hunk 9 above)
Comment (bezirg) [reply plutus-core/prelude/PlutusPrelude.hs:250]: I believe you that it does not have good performance. I am not currently relying on its performance, just its convenience for my testing. I can move it away from PlutusPrelude; let me know

File: plutus-core/plutus-core/src/PlutusCore/Core/Type.hs (hunk 7 above)
Comment (bezirg) [reply plutus-core/plutus-core/src/PlutusCore/Core/Type.hs:249]: I see. Still I am pondering between using your optimized fmap or just inline the coerce itself. It's just that `coerce` bit us before, so I was very reluctant to use it again. 

Note to self : `coerce` without type applications/signatures is dangerous!

**Comment by bezirg** [comment]
Lol, it shows if as even slower

File: plutus-core/plutus-core/stdlib/PlutusCore/StdLib/Data/Bool.hs (hunk 8 above)
Comment (effectfully) [reply plutus-core/plutus-core/stdlib/PlutusCore/StdLib/Data/Bool.hs:25]: Eek, that's a very valid reason to introduce the changes. OK, let's keep them and I'll deal with the git conflicts and the problem that it's going to be hard to unhardcode `Type` from the new machinery. Thanks for motivating me them!

File: plutus-core/plutus-core/src/PlutusCore/Core/Type.hs (hunk 7 above)
Comment (bezirg) [reply plutus-core/plutus-core/src/PlutusCore/Core/Type.hs]: What do you think about still deriving stock Functor but then using `coerceVia fmap`?

Like:

```
    size = size . coerceVia fmap fromFake
    encode = encode . coerceVia fmap fromFake
    decode =  coerceVia fmap toFake <$> decode
```

Comment (effectfully) [reply plutus-core/plutus-core/src/PlutusCore/Core/Type.hs:249]: You just derive `Functor` via `Identity` and that is simpler than calling `coerceVia` at the call site.

But I don't mind using `coerceVia` there or just `fmap` directly, performance should be identical.

Comment (bezirg) [reply plutus-core/plutus-core/src/PlutusCore/Core/Type.hs:249]: I just saw the thread, I will do that.

**Comment by effectfully** [comment]
How the actual fuck is this even possible. I'm going to take another look, although it's probably the benchmarks that are the problem here.

**Comment by effectfully** [comment]
Ship it. I rebased on `master`, squashed the commits and rerun the benchmarks, it's all [OK](https://github.com/input-output-hk/plutus/pull/5407#issuecomment-1612141306).
